### PR TITLE
AV-209062: Wildcard fqdn support in Gateway

### PR DESCRIPTION
**Pull request by akshayhavile**
TODO:
1. write down function for hostname at gateway and httproute empty --> Done
2. Write down UT for graphlayer. --> Done
3. Deletion functionality for HTTP Route mapped with Parent VS --> Done

### CODE REVIEW

**Comment by pkoshtavmware** [comment]
UnitTest files might be conflicted once https://github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pull/1468 is merged.

File: tests/gatewayapitests/ingestion/httproute_test.go [hunk 1]
```diff
@@ -174,3 +174,115 @@ func TestHTTPRouteWithBackendRefFilters(t *testing.T) {
 	akogatewayapitests.TeardownGatewayClass(t, gatewayClassName)
 	waitAndverify(t, gwClassKey)
 }
+func TestHTTPRouteGatewayWithEmptyHostnameInGateway(t *testing.T) {
+	gatewayClassName := "gateway-class-04"
+	gatewayName := "gateway-04"
```
Comment (pkoshtavmware) [inline tests/gatewayapitests/ingestion/httproute_test.go:179]: Please keep the gateway/gatewayclass/httprouteNames unique within one test file

File: ako-gateway-api/k8s/validator.go [hunk 2]
```diff
@@ -166,8 +157,8 @@ func isValidListener(key string, gateway *gatewayv1.Gateway, gatewayStatus *gate
 	for _, gatewayInNamespace := range gatewayNsList {
 		if gateway.Name != gatewayInNamespace.Name {
 			for _, gwListener := range gatewayInNamespace.Spec.Listeners {
-				if *listener.Hostname == *gwListener.Hostname || isRegexMatch(string(*listener.Hostname), string(*gwListener.Hostname), key) || isRegexMatch(string(*gwListener.Hostname), string(*listener.Hostname), key) {
-					utils.AviLog.Errorf("key: %s, msg: Hostname overlaps or is same as an existing gateway %s hostname %s", key, gatewayInNamespace.Name, *gwListener.Hostname)
+				if *listener.Hostname == *gwListener.Hostname {
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:160]: We should remove this condition alltogether

Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go:160]: I didn't get intention of removing it. If other gateway as the same hostname, we shouldn't process that gateway as same fqdn can't be associated with different vips.

File: ako-gateway-api/lib/lib.go [hunk 3]
```diff
@@ -67,6 +79,13 @@ func GetDataScriptName() string {
 	return lib.Encode(name, lib.EVHVS)
 }
 
+func GetHTTPRuleName(parentNs, parentName, routeNs, routeName, matchName string) string {
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/lib/lib.go:82]: I think I should modify this function. As empty matchname will result in creating default PG in parent VS and not http ps. This function is called from a logic which is building up http ps.

File: ako-gateway-api/nodes/avi_model_l7_translator.go [hunk 4]
```diff
@@ -40,6 +48,33 @@ func (o *AviObjectGraph) ProcessL7Routes(key string, routeModel RouteModel, pare
...
+
+	parentNode := o.GetAviEvhVS()
+	parentNs, _, parentName := lib.ExtractTypeNameNamespace(parentNsName)
+
+	// TODO(Akshay): with empty hostname at httproute, there will be no hostname attached. So need to fetch non wildcard fqdns from gw listeners
+
+	//akogatewayapiobjects.GatewayApiLister().Get
+	for _, rule := range rules {
+		// Each rule has to be converted to one httpPS
+		httpsName := akogatewayapilib.GetChildName(parentNs, parentName, routeModel.GetNamespace(), routeModel.GetName(), utils.Stringify(rule.Matches)+utils.Stringify(rule.Filters))
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/nodes/avi_model_l7_translator.go]: Done

File: ako-gateway-api/nodes/avi_model_l7_translator.go [hunk 5]
```diff
@@ -40,6 +48,33 @@ func (o *AviObjectGraph) ProcessL7Routes(key string, routeModel RouteModel, pare
...
+		}
+		if len(rule.Matches) != 0 {
+			// build HTTPPS, PG and pools
+			o.BuildParentHTTPS(key, parentNsName, parentNode[0], routeModel, rule, 0, httpsName)
+		}
+
+		if len(rule.Matches) == 0 && len(rule.Backends) != 0 {
+			// Default PG. Empty match name
+			o.BuildDefaultPGPoolForParentVS(key, parentNsName, "", parentNode[0], routeModel, rule)
+		}
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:85]: HTPPRoute
1. Can have only filter --> this will be handled by BuildHTTPPolicySet
2. Can have matches and backeend --> handled by o.BuildParentHTTPS
3. Can have filter and backend -->  so filter will be handled by BuildHTTPPolicySet and backend will be handled by o.BuildDefaultPGPoolForParentVS
4. Can have backend: Default PG with o.BuildDefaultPGPoolForParentVS
5. Empty backend: i couldn't think of use case.

File: ako-gateway-api/nodes/gateway_model_rel.go [hunk 6]
```diff
@@ -161,8 +161,13 @@ func GatewayGetGw(namespace, name, key string) ([]string, bool) {
...
 		}
 		listeners = append(listeners, gwListener)
-		hostnames[string(listenerObj.Name)] = string(*listenerObj.Hostname)
-		gwHostnames = append(gwHostnames, string(*listenerObj.Hostname))
+		if listenerObj.Hostname != nil && string(*listenerObj.Hostname) != "" {
+			hostnames[string(listenerObj.Name)] = string(*listenerObj.Hostname)
+			gwHostnames = append(gwHostnames, string(*listenerObj.Hostname))
+		} else {
+			hostnames[string(listenerObj.Name)] = utils.WILDCARD
+			gwHostnames = append(gwHostnames, utils.WILDCARD)
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/nodes/gateway_model_rel.go:170]: handled unique hostname condition.

File: tests/gatewayapitests/ingestion/httproute_test.go (hunk 1 above)
Comment (akshayhavile) [reply tests/gatewayapitests/ingestion/httproute_test.go]: addressed

**Comment by pkoshtavmware** [comment]
Please resolve conflicts and check if all these cases will pass https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/tests/gateway-http-listener-isolation.go

File: ako-gateway-api/k8s/validator.go (hunk 2 above)
Comment (pkoshtavmware) [reply ako-gateway-api/k8s/validator.go]: Please check yaml for this conformance test https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/tests/gateway-secret-reference-grant-all-in-namespace.go. This wont pass unless this check is removed

Comment (pkoshtavmware) [reply ako-gateway-api/k8s/validator.go]: Also, if one of the listener is invalid, as per today's discussion we should partially accept the gateway and set that listener status to be unaccepted

File: ako-gateway-api/nodes/avi_model_l7_translator.go [hunk 7]
```diff
@@ -173,70 +220,169 @@ func (o *AviObjectGraph) BuildPGPool(key, parentNsName string, childVsNode *node
...
+			// Replace the poolNode.
+			parentVsNode.ReplaceEvhPoolInEVHNode(poolNode, key)
+			// Add pool to list of Pools attached to Parent VS
+			httppsPGPool.Pool = append(httppsPGPool.Pool, poolNode.Name)
+			uniquePools := sets.NewString(httppsPGPool.Pool...)
+			httppsPGPool.Pool = uniquePools.List()
+		}
 		pool_ref := fmt.Sprintf("/api/pool?name=%s", poolNode.Name)
-		ratio := uint32(httpbackend.Backend.Weight)
+		o.BuildBackendFiltersModel(key, poolName, backend, parentVsNode)
```
Comment (pkoshtavmware) [inline ako-gateway-api/nodes/avi_model_l7_translator.go:370]: Please check if backendRefFilters flow is working as expected in this case as well since in the former case HTTPRoute with hostname we attach the datascript to the childVs, however here we add it to the parentVs

File: ako-gateway-api/nodes/avi_model_l7_translator.go [hunk 8]
```diff
@@ -309,30 +455,132 @@ func (o *AviObjectGraph) BuildVHMatch(key string, parentNsName string, routeType
...
+	}
+	if req_rule == nil {
+		req_rule = &models.HTTPRequestRule{Name: &httpPSName, Enable: proto.Bool(true)}
+		vsNode.HttpPolicyRefs[index].RequestRules = append(vsNode.HttpPolicyRefs[index].RequestRules, req_rule)
+
+		rule_index = len(vsNode.HttpPolicyRefs[index].RequestRules) - 1
+	}
+	parentNs, _, parentName := lib.ExtractTypeNameNamespace(parentNsName)
+
+	//port match is not required as its empty hostname at httproute. So it will map to all listener ports of Parent
```
Comment (pkoshtavmware) [inline ako-gateway-api/nodes/avi_model_l7_translator.go:496]: Even if SectionName is specified ?

File: ako-gateway-api/nodes/dequeue_ingestion.go [hunk 9]
```diff
@@ -212,29 +212,73 @@ func saveAviModel(modelName string, aviGraph *nodes.AviObjectGraph, key string)
...
+		ok := saveAviModel(modelName, o.AviObjectGraph, key)
+		if ok && len(o.AviObjectGraph.GetOrderedNodes()) != 0 && !fullsync {
+			sharedQueue := utils.SharedWorkQueue().GetQueueByName(utils.GraphLayer)
+			nodes.PublishKeyToRestLayer(modelName, key, sharedQueue)
+		}
+	} else {
+		// check parent association
+		found, localHTTPPSPGPools := akogatewayapiobjects.GatewayApiLister().GetGatewayRouteToHTTPSPGPool(parentNsName + "/" + routeModel.GetType() + "/" + routeModel.GetNamespace() + "/" + routeModel.GetName())
+		if !found {
+			utils.AviLog.Warnf("key: %s, msg: no child vs mapped to this route %s", key, routeModel.GetName())
```
Comment (pkoshtavmware) [inline ako-gateway-api/nodes/dequeue_ingestion.go:239]: Might want to changes log message here

File: ako-gateway-api/nodes/gateway_model_rel.go [hunk 10]
```diff
@@ -259,17 +267,22 @@ func HTTPRouteToGateway(namespace, name, key string) ([]string, bool) {
...
+
 					hostnameMatched := false
 					for _, routeHostname := range hrObj.Spec.Hostnames {
-						if strings.HasSuffix(string(routeHostname), listenerHostname) && akogatewayapilib.VerifyHostnameSubdomainMatch(string(routeHostname)) {
-							hostnameIntersection = append(hostnameIntersection, string(routeHostname))
-							hostnameMatched = true
+						// When Gateway hostname is empty, then just check validity of hostname and append it.
+						// When hostname in HTTProute has wildcard
+						// When there is exact match
+						if strings.HasSuffix(string(routeHostname), listenerHostname) || listenerHostname == "" || strings.HasPrefix(string(routeHostname), utils.WILDCARD) {
```
Comment (pkoshtavmware) [inline ako-gateway-api/nodes/gateway_model_rel.go:278]: Will this condition not lead to true even if listener hostname is a.b.c and routehostname is d.a.b.c ?
both of these hostnames are neither overlapping nor exact match

File: ako-gateway-api/nodes/gateway_model_rel.go [hunk 11]
```diff
@@ -259,17 +265,23 @@ func HTTPRouteToGateway(namespace, name, key string) ([]string, bool) {
...
+						if strings.HasSuffix(string(routeHostname), listenerHostname) || listenerHostname == "" || strings.HasPrefix(string(routeHostname), "*") {
+							// Here i need to check if route hostname is empty
+							if akogatewayapilib.VerifyHostnameSubdomainMatch(string(routeHostname)) {
+								hostnameIntersection = append(hostnameIntersection, string(routeHostname))
+								hostnameMatched = true
+							}
 						}
 					}
-					if hostnameMatched && !utils.HasElem(listenerList, listener) {
+					if (hostnameMatched && !utils.HasElem(listenerList, listener)) || len(hrObj.Spec.Hostnames) == 0 {
```
[the remark this answers is not in the document]
Comment (pkoshtavmware) [reply ako-gateway-api/nodes/gateway_model_rel.go:285]: Even if sectionName is specified in the parentRef ?

File: ako-gateway-api/k8s/validator.go [hunk 12]
```diff
@@ -378,27 +366,67 @@ func validateParentReference(key string, httpRoute *gatewayv1.HTTPRoute, httpRou
...
+				if strings.HasPrefix(string(host), utils.WILDCARD) {
+					isHttpRouteHostFqdnWildcard = true
+				}
+				if isHttpRouteHostFqdnWildcard && isListnerFqdnWildcard {
+					// both are true. Match nonwildcard part
+					nonWildCardHttpRouteHostname := strings.Split(string(host), utils.WILDCARD)
+					nonWildCardGWHostname := strings.Split(string(*hostInListener), utils.WILDCARD)
+					// Use case: 1. GW: *.avi.internal HttpRoute: *.bar.avi.internal
+					// USe case: 2. GW: *.bar.avi.internal HttpRoute: *.avi.internal
+					if strings.Contains(nonWildCardHttpRouteHostname[1], nonWildCardGWHostname[1]) || strings.Contains(nonWildCardGWHostname[1], nonWildCardHttpRouteHostname[1]) {
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:403]: It will even return true for *.avi.internal.abc and *.avi.internal

File: ako-gateway-api/k8s/validator.go [hunk 13]
```diff
@@ -378,27 +366,67 @@ func validateParentReference(key string, httpRoute *gatewayv1.HTTPRoute, httpRou
...
-			continue
-		}
-		var matched bool
-		for _, host := range httpRoute.Spec.Hostnames {
-			matched = matched || expr.MatchString(string(host))
+				}
+
+			}
+			// if there are no hostnames specified, all parent listneres should be matched.
+			if len(httpRoute.Spec.Hostnames) == 0 {
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:427]: What if SectionName is specified

File: ako-gateway-api/k8s/validator.go (hunk 2 above)
Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go]: For this change, we need to merge the gateways. This will major change as it will involve
1. Uniquely identifying all valid listeners, clubbing those together. 
2. Associating those gateway with single Parent VS. So it will introduce naming convention changes.
3. Associating routes with common parent VS.

We can take this up as part of another PR. Please raise Jira for that.

File: ako-gateway-api/nodes/avi_model_l7_translator.go [hunk 14]
```diff
@@ -173,70 +220,169 @@ func (o *AviObjectGraph) BuildPGPool(key, parentNsName string, childVsNode *node
...
+	parentNs, _, parentName := lib.ExtractTypeNameNamespace(parentNsName)
+
+	PGName := akogatewayapilib.GetPoolGroupName(parentNs, parentName,
+		httpRouteNamespace, httpRouteName, matchName)
+
+	// Check default PG name is same as that of already assigned
+	if matchName == "" && parentVsNode.DefaultPoolGroup != "" && parentVsNode.DefaultPoolGroup != PGName {
+		utils.AviLog.Warnf("key: %s, msg: Parent VS already has default PG. HttpRoute %s/%s is not attached to Gateway %s", key, httpRouteNamespace, httpRouteName, parentNsName)
+		//TODO(Akshay): add condition here.
+		return pgAttachedToVS
```
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:308]: Naming convention of PG involves GW namespace, GW name, HTTP Route namespace, HTTP Route name, match name.  So if match name is empty, that means only BackendRef exists, so as per your comment, your are saying there will be 2 rules in HTTP Route without any match and only backendref mentioned. What is the use case for this of sending default traffic to two different pools? Can you please explain more on this?

File: ako-gateway-api/nodes/avi_model_l7_translator.go (hunk 7 above)
Comment (pkoshtavmware) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:372]: Is this validated ?

File: ako-gateway-api/nodes/avi_model_l7_translator.go [hunk 15]
```diff
@@ -75,7 +122,7 @@ func (o *AviObjectGraph) BuildChildVS(key string, routeModel RouteModel, parentN
 		Host:        hosts,
 	}
 	for _, host := range hosts {
-		if !utils.HasElem(parentNode[0].VSVIPRefs[0].FQDNs, host) {
+		if !strings.Contains(host, utils.WILDCARD) && !utils.HasElem(parentNode[0].VSVIPRefs[0].FQDNs, host) {
```
[the remark this answers is not in the document]
Comment (pkoshtavmware) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:125]: But in that case, even if somebody with DNS other than Avi, tries to create the gateway with wildcard in listener hostname, we will still not add that to vsvip. Will the traffic go through for them ?
If not then whole purpose of supporting wildcard in gateway listener hostname is defeated

File: ako-gateway-api/k8s/validator.go (hunk 2 above)
Comment (pkoshtavmware) [reply ako-gateway-api/k8s/validator.go]: https://avinetworks.atlassian.net/browse/AV-212611

File: ako-gateway-api/nodes/avi_model_l7_translator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pkoshtavmware) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:85]: Could you please add that validation?
As of today, neither K8s nor AKO has that validation in place and the route->rule does not gets rejected

File: ako-gateway-api/lib/lib.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pkoshtavmware) [reply ako-gateway-api/lib/lib.go:82]: This case is setting httproute as invalid and rejecting it. I have tested this out just now. So it should be fine

File: ako-gateway-api/k8s/validator.go [hunk 16]
```diff
@@ -166,8 +157,11 @@ func isValidListener(key string, gateway *gatewayv1.Gateway, gatewayStatus *gate
...
 			for _, gwListener := range gatewayInNamespace.Spec.Listeners {
-				if *listener.Hostname == *gwListener.Hostname || isRegexMatch(string(*listener.Hostname), string(*gwListener.Hostname), key) || isRegexMatch(string(*gwListener.Hostname), string(*listener.Hostname), key) {
-					utils.AviLog.Errorf("key: %s, msg: Hostname overlaps or is same as an existing gateway %s hostname %s", key, gatewayInNamespace.Name, *gwListener.Hostname)
+				if gwListener.Hostname == nil {
+					continue
+				}
+				if listener.Hostname != nil && *listener.Hostname == *gwListener.Hostname {
+					utils.AviLog.Errorf("key: %s, msg: Hostname is same as an existing gateway %s hostname %s", key, gatewayInNamespace.Name, *gwListener.Hostname)
 					defaultCondition.
 						Message("Hostname overlaps or is same as an existing gateway hostname").
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:166]: We can change this message now as it only error out when there is an exact match and no overlapping

File: ako-gateway-api/k8s/validator.go [hunk 17]
```diff
@@ -166,8 +157,11 @@ func isValidListener(key string, gateway *gatewayv1.Gateway, gatewayStatus *gate
 	for _, gatewayInNamespace := range gatewayNsList {
 		if gateway.Name != gatewayInNamespace.Name {
 			for _, gwListener := range gatewayInNamespace.Spec.Listeners {
-				if *listener.Hostname == *gwListener.Hostname || isRegexMatch(string(*listener.Hostname), string(*gwListener.Hostname), key) || isRegexMatch(string(*gwListener.Hostname), string(*listener.Hostname), key) {
-					utils.AviLog.Errorf("key: %s, msg: Hostname overlaps or is same as an existing gateway %s hostname %s", key, gatewayInNamespace.Name, *gwListener.Hostname)
+				if gwListener.Hostname == nil {
+					continue
+				}
+				if listener.Hostname != nil && *listener.Hostname == *gwListener.Hostname {
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:163]: Also if we are planning to keep this check forever, due to limitations on controller side, can we use `gatewayToHostnameStore` here to list hostnames associated with a gateway instead of looping through gateway hostnames for consistency?

Any thoughts ?
@akshayhavile @aaha97

File: ako-gateway-api/k8s/validator.go [hunk 18]
```diff
@@ -166,8 +157,11 @@ func isValidListener(key string, gateway *gatewayv1.Gateway, gatewayStatus *gate
...
+					utils.AviLog.Errorf("key: %s, msg: Hostname is same as an existing gateway %s hostname %s", key, gatewayInNamespace.Name, *gwListener.Hostname)
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:164]: We can change this message here as well as in logs now as it only errors out when there is an exact match and no overlapping

File: ako-gateway-api/k8s/validator.go [hunk 19]
```diff
@@ -378,27 +366,67 @@ func validateParentReference(key string, httpRoute *gatewayv1.HTTPRoute, httpRou
...
+				isHttpRouteHostFqdnWildcard := false
+				if strings.HasPrefix(string(host), utils.WILDCARD) {
+					isHttpRouteHostFqdnWildcard = true
+				}
+				if isHttpRouteHostFqdnWildcard && isListenerFqdnWildcard {
+					// both are true. Match nonwildcard part
+					nonWildCardHttpRouteHostname := strings.Split(string(host), utils.WILDCARD)
+					nonWildCardGWHostname := strings.Split(string(*hostInListener), utils.WILDCARD)
+					// Use case: 1. GW: *.avi.internal HttpRoute: *.bar.avi.internal
+					// USe case: 2. GW: *.bar.avi.internal HttpRoute: *.avi.internal
```
Comment (pkoshtavmware) [inline ako-gateway-api/k8s/validator.go:402]: Is this splitting and sending only nonWildCard part of hostnames to CheckSubdomainOverlapping even required ?
As far as I understand, CheckSubdomainOverlapping itself can handle that

File: ako-gateway-api/nodes/avi_model_l7_translator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:85]: @DixitAakash: HTTPRoutes, attached to same Parent Ref, with different backend refs and no other section, implies configuration will lead to 2 default actions which is wrong configuration.

File: ako-gateway-api/nodes/avi_model_l7_translator.go (hunk 7 above)
Comment (akshayhavile) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:372]: bakendref changes are on hold so not tested and not enabled it also.

File: ako-gateway-api/k8s/validator.go (hunk 17 above)
Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go:163]: @pkoshtavmware : This change has to be taken separately. Reasons:
1. `gatwewayToHostnameStore` stores hostnames associated with gateway. Currently, this is done during dequeueingestion phase which is always called after validation of gateway done with other gateway hostnames.
2. We can store these values in this data structure upfront in validation phase.
3. This will also require another datastructure `hostsToGateway` list to maintain hostname to Gateways.
4. So when any gateway gets deleted or listeners hostname changes, we need to fetch gateway associated with that hostname. Remove entry of deleted or changed gateway and then trigger add/update for other gateway.
5. So this will create of flow of deletion, addition of parents, children virtualservice. 
6. During bootup, we can arrange gateway in creation timestamp basis(which we do today also but after validaton) and then update this datastructure so no need to maintain anything at Controller side about gateway <-> hostname relations.

File: ako-gateway-api/nodes/avi_model_l7_translator.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:125]: Finding out proper IP address for Given hostname is work of DNS. So even though , implementation is not adding wildcard to vsvip, child vs is capable of handling wildcard requests. So if requests land on Parent VS, then it will be served by appropriate child VS.

File: ako-gateway-api/nodes/avi_model_l7_translator.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (akshayhavile) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:310]: when there are two HTTP Routes, attached to same Gateway, without any rules or filters and only backendref.

File: ako-gateway-api/k8s/validator.go (hunk 18 above)
Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go:164]: done

File: ako-gateway-api/k8s/validator.go (hunk 19 above)
Comment (akshayhavile) [reply ako-gateway-api/k8s/validator.go:400]: done,

File: ako-gateway-api/nodes/avi_model_l7_translator.go (hunk 7 above)
Comment (pkoshtavmware) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:372]: okay

**Review by pkoshtavmware — APPROVED** [review]
LGTM

**Comment by akshayhavile** [comment]
build ako

**Review by DixitAakash — APPROVED** [review]

**Comment by akshayhavile** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

File: ako-gateway-api/nodes/avi_model_l7_translator.go [hunk 20]
```diff
@@ -599,6 +599,7 @@ func (o *AviObjectGraph) BuildHTTPPolicySet(key string, vsNode *nodes.AviEvhVsNo
 
 func (o *AviObjectGraph) BuildHTTPPolicySetHTTPRequestRules(key, httpPSName string, vsNode *nodes.AviEvhVsNode, routeModel RouteModel, filters []*Filter, index int) {
 	requestRule := &models.HTTPRequestRule{Name: &httpPSName, Enable: proto.Bool(true), Index: proto.Int32(int32(index + 1))}
+	vsNode.HttpPolicyRefs[index].RequestRules = []*models.HTTPRequestRule{}
```
Comment (pkoshtavmware) [inline ako-gateway-api/nodes/avi_model_l7_translator.go:602]: I did not understand why this change is needed

Comment (akshayhavile) [reply ako-gateway-api/nodes/avi_model_l7_translator.go:602]: This is to handle update case on filter section of httpRoute

**Review by pkoshtavmware — APPROVED** [review]
LGTM

**Review by DixitAakash — APPROVED** [review]

**Comment by akshayhavile** [comment]
***** Local UT Results****
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**k8stest**       30.390s coverage: 8.8% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**integrationtest**       484.021s        coverage: 26.2% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**ingresstests**  979.847s        coverage: 38.1% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**evhtests**      1332.209s       coverage: 37.4% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**evhtests**      1237.663s       coverage: 36.9% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**dedicatedevhtests**     338.407s        coverage: 27.0% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**dedicatedevhtests**     434.389s        coverage: 27.4% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**oshiftroutetests**      1267.932s       coverage: 32.1% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**bootuptests**   20.191s coverage: 4.3% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**multicloudtests**       170.447s        coverage: 8.2% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**advl4tests**    46.707s coverage: 19.0% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**namespacesynctests**    152.376s        coverage: 21.1% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**servicesapitests**      78.997s coverage: 22.3% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**npltests**      116.794s        coverage: 23.7% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**dedicatedvstests**      154.420s        coverage: 28.9% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**hatests**       33.881s coverage: 24.9% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**cnitests**      103.550s        coverage: 13.1% of statements in ./...
 PASS  Test avi-secret creation when avi controller credentials are specified or not specified in values.yaml   tests/avi-secret_test.yaml
 PASS  Test AKO's ClusterRole with GatewayAPI   tests/clusterrole_test.yaml
 PASS  Test GatewayClass creation with GatewayAPI       tests/gatewayclass_test.yaml
 PASS  Test AKO's StatefulSet with GatewayAPI   tests/statefulset_test.yaml

Charts:      1 passed, 1 total
Test Suites: 4 passed, 4 total
Tests:       9 passed, 9 total
Snapshot:    0 passed, 0 total
Time:        54.065735ms

ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**cnitests**      27.285s coverage: 12.8% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**infratests**    0.932s
sudo docker run \
-w=/go/src/github.com/vmware/load-balancer-and-ingress-services-for-kubernetes \
-v /var/work/load-balancer-and-ingress-services-for-kubernetes:/go/src/github.com/vmware/load-balancer-and-ingress-services-for-kubernetes golang:bullseye \
go test -mod=vendor github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**multitenancytests**/... -failfast -timeout 0 -coverprofile cover-21.out -coverpkg=./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/multitenancytests     70.610s coverage: 34.1% of statements in ./...
sudo docker run \
-w=/go/src/github.com/vmware/load-balancer-and-ingress-services-for-kubernetes \
-v /var/work/load-balancer-and-ingress-services-for-kubernetes:/go/src/github.com/vmware/load-balancer-and-ingress-services-for-kubernetes golang:bullseye \
go test -mod=vendor github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**gatewayapitests**/... -failfast -timeout 0 -coverprofile cover-20.out -coverpkg=./...
        github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/gatewayapitests               coverage: 0.0% of statements
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**gatewayapitests/graphlayer**    258.840s        coverage: 18.9% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**gatewayapitests/ingestion**     270.645s        coverage: 3.0% of statements in ./...
ok      github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/**gatewayapitests/status**        314.736s        coverage: 16.0% of statements in ./...
